### PR TITLE
8289765: JDI EventSet/resume/resume008 failed with "ERROR: suspendCounts don't match for : VirtualThread-unparker"

### PR DESCRIPTION
**Pull request by plummercj**
The test failure is caused by the arrival of unexpected ThreadStartEvents, which mess up the debugger side. The events are for threads we normally only see getting created when using virtual threads, such as carrier threads and the VirtualThread-unparker thread. Theoretically this issue could happen without virtual threads due to other VM threads starting up such as Common-Cleaner, but we haven't seen it fail for that reason yet.

The test is testing proper thread suspension for ThreadStartEvent using each of the 3 suspension policy types. The debuggee creates a sequence of 3 debuggee threads, each one's timing coordinated with some complicated synchronization with the debugger using breakpoints and the setting of fields in the debuggee (and careful placement of suspend/resume in the debugger). The ThreadStartRequests that the debugger sets up always use a "count filter" of 1, which means the requests are good for delivering exactly 1 ThreadStartEvent, and any that come after the first will get filtered out. So when an an unexpected ThreadStartEvent arrives for something like a carrier thread, this prematurely moves the debugger on to the next step, and the synchronization with the debuggee gets messed up.

The first step in fixing this test was to remove the count filter, so the request can handle any number of ThreadStartEvents.

The next step was then fixing the test library code in EventHandler.java so it would filter out any undesired ThreadStartEvents, so the test just ends up getting one event, and always for the thread it is expecting. There are a few parts to this. One is improving EventFilters.filter() to filter out more threads that tend to be created during VM startup, including carrier threads and the VirtualThread-unparker thread.

It was necessary to add some calls EventFilters.filter() from EventHandler. This was done by adding a ThreadStartEvent listener for the "spurious" thread starts (those the test debuggee does not create). This listener is added by waitForRequestedEventCommon(), which is indirectly called by the test when is calls waitForRequestedEventSet().

There is a also 2nd place where the ThreadStartEvent listener for "spurious" threads is needed. It is now also installed with the default listeners that are always in place. It is needed when the test is not actually waiting for a ThreadStartEvent, but is waiting for a BreakpointEvent. waitForRequestedEventCommon() is not used in this case (so none of its listeners are installed), but the default listeners are always in place and can be used to filter these ThreadStartEvents. Note this filter will also be in place when calling waitForRequestedEventCommon(), but we can't realy on it when waitForRequestedEventCommon() is used for ThreadStartEvents because the spurious ThreadStartEvent will be seen and returned before we ever get to the default filter. So we actually end up with this ThreadStartEvent listener installed twice during  waitForRequestedEventCommon().

I did a bit of cleanup on the test, mostly renaming of threads and ThreadStartRequests so they are easier to match up with the iteration # we use in both the debuggee and debugger (0, 1, and 2). The only real change in the test itself is removing the filter count, and verifying that the ThreadStartEvent is for the expected thread.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289765](https://bugs.openjdk.org/browse/JDK-8289765): JDI EventSet/resume/resume008 failed with "ERROR: suspendCounts don't match for : VirtualThread-unparker"

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [f159416c](https://git.openjdk.org/jdk/pull/12861/files/f159416c9972f14cf96d63d583df389f6173029d)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [f159416c](https://git.openjdk.org/jdk/pull/12861/files/f159416c9972f14cf96d63d583df389f6173029d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12861/head:pull/12861` \
`$ git checkout pull/12861`

Update a local copy of the PR: \
`$ git checkout pull/12861` \
`$ git pull https://git.openjdk.org/jdk pull/12861/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12861`

View PR using the GUI difftool: \
`$ git pr show -t 12861`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12861.diff">https://git.openjdk.org/jdk/pull/12861.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/vmTestbase/nsk/share/jdi/EventHandler.java [hunk 1]
```diff
@@ -251,6 +251,41 @@ private void createDefaultEventRequests() {
...
+         * This listener catches spurious thread creations that we want to ignore.
+         */
+        return (new EventListener() {
+            boolean handled = false;
+
+            public boolean eventReceived(Event event) {
+                handled = false;
+                if (event instanceof ThreadStartEvent) {
+                    if (EventFilters.filtered(event)) {
+                        display(owner +": Ignoring spurious thread creation: " + event);
```
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/share/jdi/EventHandler.java:270]: Nit: Space is needed after first '+' sign.

**Review by sspitsyn — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/EventSet/resume/resume008.java [hunk 2]
```diff
@@ -137,6 +137,17 @@ protected void testRun() {
...
+
+            // Make sure any spurious ThreadStartEvent got filtered out.
+            if (newEvent instanceof ThreadStartEvent) {
+                ThreadStartEvent t = (ThreadStartEvent)newEvent;
+                if (!t.thread().name().equals(resume008a.THREAD_NAME_PREFIX + i)) {
+                    setFailedStatus("ERROR: ThreadStartEvent is not for expected thread:  " +
+                                    t.thread().name());
+                    return;
+                }
+            }
```
Comment (kevinjwalls) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/EventSet/resume/resume008.java:150]: Not your typo here, but just below on line 157 now there's the "propety" spelling error which was in the failure logs.

**Review by kevinjwalls — APPROVED** [review]
Looks good.
If in future we ever find ourselves updating that filter list, it could use a parameter to let us require only THREAD_NAME_PREFIX, rather than maintain the list of what we don't want. 8-)

**Comment by plummercj** [comment]
> If in future we ever find ourselves updating that filter list, it could use a parameter to let us require only THREAD_NAME_PREFIX, rather than maintain the list of what we don't want. 8-)

The threads not to filter could be passed to EventHandler.waitForRequestedEventSet(), but that doesn't solve the other issue with JDIBase.breakpointForCommunication(), which doesn't know the set of threads we care about and also relies on EventFilters.filtered(event).

I'm surprised we don't have more failures due to this. I think maybe most were fixed in other ways. For example, see https://github.com/openjdk/jdk/pull/5567, which fixed a similar problem.  In that case the test uses JDIBase.getEventSetForThreadStartDeath(). I took a different approach since this test uses EventHandler.waitForRequestedEventSet(), and I didn't think converting to getEventSetForThreadStartDeath() was appropriate.

**Comment by plummercj** [comment]
Thanks Kevin and Serguei!

/integrate

**Pull request closed without merging**